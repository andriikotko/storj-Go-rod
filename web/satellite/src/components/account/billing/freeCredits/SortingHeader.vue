// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="sort-header-container">
        <div class="sort-header-container__item">
            <p class="sort-header-container__item__name">CREDIT TYPE</p>
        </div>
        <div class="sort-header-container__item">
            <p class="sort-header-container__item__name">EXPIRATION</p>
        </div>
        <div class="sort-header-container__item">
            <p class="sort-header-container__item__name">EARNED AMOUNT</p>
        </div>
        <div class="sort-header-container__item available">
            <p class="sort-header-container__item__name">AVAILABLE</p>
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

@Component
export default class SortingHeader extends Vue {}
</script>

<style scoped lang="scss">
    .sort-header-container {
        display: flex;
        width: 100%;

        &__item {
            text-align: left;
            min-width: 28%;

            &__name {
                font-family: 'font_regular', sans-serif;
                font-weight: 500;
                font-size: 14px;
                line-height: 19px;
                letter-spacing: 0.03em;
                color: #adadad;
                margin: 0;
            }
        }
    }

    .available {
        min-width: 16%;
        text-align: right;
    }
</style>
