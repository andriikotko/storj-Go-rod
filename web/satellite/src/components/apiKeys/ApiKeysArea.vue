// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="api-keys-area">
        <h1 class="api-keys-area__title" v-if="!isEmpty">API Keys</h1>
        <div class="api-keys-area__container">
            <ApiKeysCreationPopup
                @closePopup="closeNewApiKeyPopup"
                @showCopyPopup="showCopyApiKeyPopup"
                :is-popup-shown="isNewApiKeyPopupShown"
            />
            <ApiKeysCopyPopup
                :is-popup-shown="isCopyApiKeyPopupShown"
                :api-key-secret="apiKeySecret"
                @closePopup="closeCopyNewApiKeyPopup"
            />
            <div v-if="isHeaderShown" class="api-keys-header">
                <VHeader
                    ref="headerComponent"
                    placeholder="API Key"
                    :search="onSearchQueryCallback">
                    <div class="header-default-state" v-if="isDefaultHeaderState">
                        <VButton
                            class="button"
                            label="+ Create API Key"
                            width="180px"
                            height="48px"
                            :on-press="onCreateApiKeyClick"
                        />
                    </div>
                    <div class="header-selected-api-keys" v-if="areApiKeysSelected">
                        <span class="header-selected-api-keys__confirmation-label" v-if="isDeleteClicked">
                            Are you sure you want to delete <b>{{selectedAPIKeysCount}}</b> {{apiKeyCountTitle}} ?
                        </span>
                        <div class="header-selected-api-keys__buttons-area">
                            <VButton
                                class="button deletion"
                                label="Delete"
                                width="122px"
                                height="48px"
                                :on-press="onDeleteClick"
                            />
                            <VButton
                                class="button"
                                label="Cancel"
                                width="122px"
                                height="48px"
                                is-transparent="true"
                                :on-press="onClearSelection"
                            />
                            <span class="header-selected-api-keys__info-text" v-if="!isDeleteClicked">
                                <b>{{selectedAPIKeysCount}}</b> API Keys selected
                            </span>
                        </div>
                    </div>
                </VHeader>
                <div class="blur-content" v-if="isDeleteClicked"></div>
                <div class="blur-search" v-if="isDeleteClicked"></div>
            </div>
            <div v-if="!isEmpty" class="api-keys-items">
                <SortingHeader :on-header-click-callback="onHeaderSectionClickCallback"/>
                <div class="api-keys-items__content">
                    <VList
                        :data-set="apiKeyList"
                        :item-component="itemComponent"
                        :on-item-click="toggleSelection"
                    />
                </div>
                <VPagination
                    v-if="totalPageCount > 1"
                    class="pagination-area"
                    ref="pagination"
                    :total-page-count="totalPageCount"
                    :on-page-click-callback="onPageClick"
                />
            </div>
            <div class="empty-search-result-area" v-if="isEmptySearchResultShown">
                <h1 class="empty-search-result-area__title">No results found</h1>
                <EmptySearchResultIcon class="empty-search-result-area__image"/>
            </div>
            <NoApiKeysArea
                :on-button-click="onCreateApiKeyClick"
                v-if="isEmptyStateShown"
            />
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import ApiKeysItem from '@/components/apiKeys/ApiKeysItem.vue';
import NoApiKeysArea from '@/components/apiKeys/NoApiKeysArea.vue';
import SortingHeader from '@/components/apiKeys/SortingHeader.vue';
import VButton from '@/components/common/VButton.vue';
import VHeader from '@/components/common/VHeader.vue';
import VList from '@/components/common/VList.vue';
import VPagination from '@/components/common/VPagination.vue';

import EmptySearchResultIcon from '@/../static/images/common/emptySearchResult.svg';

import { API_KEYS_ACTIONS } from '@/store/modules/apiKeys';
import { ApiKey, ApiKeyOrderBy } from '@/types/apiKeys';
import { SortDirection } from '@/types/common';
import { SegmentEvent } from '@/utils/constants/analyticsEventNames';

import ApiKeysCopyPopup from './ApiKeysCopyPopup.vue';
import ApiKeysCreationPopup from './ApiKeysCreationPopup.vue';

// header state depends on api key selection state
/**
 * HeaderState is enumerable of page's header states.
 * Depends on api key selection state.
 */
enum HeaderState {
    DEFAULT = 0,
    ON_SELECT,
}

const {
    FETCH,
    DELETE,
    TOGGLE_SELECTION,
    CLEAR,
    CLEAR_SELECTION,
    SET_SEARCH_QUERY,
    SET_SORT_BY,
    SET_SORT_DIRECTION,
} = API_KEYS_ACTIONS;

declare interface ResetPagination {
    resetPageIndex(): void;
}

@Component({
    components: {
        NoApiKeysArea,
        VList,
        VHeader,
        ApiKeysItem,
        VButton,
        ApiKeysCreationPopup,
        ApiKeysCopyPopup,
        VPagination,
        SortingHeader,
        EmptySearchResultIcon,
    },
})
export default class ApiKeysArea extends Vue {
    private FIRST_PAGE = 1;
    /**
     * Indicates if delete confirmation state should appear.
     */
    private isDeleteClicked: boolean = false;
    /**
     * Indicates if api key name input state should appear.
     */
    private isNewApiKeyPopupShown: boolean = false;
    /**
     * Indicates if copy api key state should appear.
     * Should only appear once
     */
    private isCopyApiKeyPopupShown: boolean = false;
    private apiKeySecret: string = '';

    public $refs!: {
        pagination: HTMLElement & ResetPagination;
    };

    /**
     * Lifecycle hook after initial render where list of existing api keys is fetched.
     */
    public async mounted(): Promise<void> {
        await this.$store.dispatch(FETCH, 1);
        this.$segment.track(SegmentEvent.API_KEYS_VIEWED, {
            project_id: this.$store.getters.selectedProject.id,
            api_keys_count: this.selectedAPIKeysCount,
        });
    }

    /**
     * Lifecycle hook before component destruction.
     * Clears existing api keys selection and search.
     */
    public async beforeDestroy(): Promise<void> {
        this.onClearSelection();
        await this.$store.dispatch(SET_SEARCH_QUERY, '');
    }

    /**
     * Toggles api key selection.
     * @param apiKey
     */
    public async toggleSelection(apiKey: ApiKey): Promise<void> {
        await this.$store.dispatch(TOGGLE_SELECTION, apiKey);
    }

    /**
     * Starts creating API key process.
     * Makes Create API Key popup visible.
     */
    public onCreateApiKeyClick(): void {
        this.isNewApiKeyPopupShown = true;
    }

    /**
     * Holds on button click login for deleting API key process.
     */
    public onDeleteClick(): void {
        if (!this.isDeleteClicked) {
            this.isDeleteClicked = true;

            return;
        }

        this.delete();
    }

    /**
     * Clears API Keys selection.
     */
    public onClearSelection(): void {
        this.$store.dispatch(CLEAR_SELECTION);
        this.isDeleteClicked = false;
    }

    /**
     * Closes Create API Key popup.
     */
    public closeNewApiKeyPopup(): void {
        this.isNewApiKeyPopupShown = false;
    }

    /**
     * Closes Create API Key popup.
     */
    public showCopyApiKeyPopup(secret: string): void {
        this.isCopyApiKeyPopupShown = true;
        this.apiKeySecret = secret;
    }

    /**
     * Makes Copy API Key popup visible
     */
    public closeCopyNewApiKeyPopup(): void {
        this.isCopyApiKeyPopupShown = false;
    }

    /**
     * Deletes selected api keys, fetches updated list and changes area state to default.
     */
    private async delete(): Promise<void> {
        try {
            await this.$store.dispatch(DELETE);
            await this.$notify.success(`API keys deleted successfully`);
            this.$segment.track(SegmentEvent.API_KEY_DELETED, {
                project_id: this.$store.getters.selectedProject.id,
            });
        } catch (error) {
            await this.$notify.error(error.message);
        }

        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        this.isDeleteClicked = false;

        if (this.totalPageCount > 1) {
            this.$refs.pagination.resetPageIndex();
        }
    }

    /**
     * Returns API Key item component.
     */
    public get itemComponent() {
        return ApiKeysItem;
    }

    /**
     * Returns api keys from store.
     */
    public get apiKeyList(): ApiKey[] {
        return this.$store.getters.apiKeys;
    }

    /**
     * Returns api keys pages count from store.
     */
    public get totalPageCount(): number {
        return this.$store.state.apiKeysModule.page.pageCount;
    }

    /**
     * Returns api keys label depends on api keys count.
     */
    public get apiKeyCountTitle(): string {
        return this.selectedAPIKeysCount === 1 ? 'api key' : 'api keys';
    }

    /**
     * Indicates if no api keys in store.
     */
    public get isEmpty(): boolean {
        return this.$store.getters.apiKeys.length === 0;
    }

    /**
     * Indicates if there is search query in store.
     */
    public get hasSearchQuery(): boolean {
        return this.$store.state.apiKeysModule.cursor.search;
    }

    /**
     * Returns amount of selected API Keys from store.
     */
    public get selectedAPIKeysCount(): number {
        return this.$store.state.apiKeysModule.selectedApiKeysIds.length;
    }

    /**
     * Returns page's header state depending on selected API Keys amount.
     */
    public get headerState(): number {
        return this.selectedAPIKeysCount > 0 ? HeaderState.ON_SELECT : HeaderState.DEFAULT;
    }

    /**
     * Indicates if page's header is shown.
     */
    public get isHeaderShown(): boolean {
        return !this.isEmpty || this.hasSearchQuery;
    }

    /**
     * Indicates if page's header is in default state.
     */
    public get isDefaultHeaderState(): boolean {
        return this.headerState === HeaderState.DEFAULT;
    }

    /**
     * Indicates if page's header is in selected state.
     */
    public get areApiKeysSelected(): boolean {
        return this.headerState === HeaderState.ON_SELECT;
    }

    /**
     * Indicates if page is in empty search result state.
     */
    public get isEmptySearchResultShown(): boolean {
        return this.isEmpty && this.hasSearchQuery;
    }

    /**
     * Indicates if page is in empty state.
     */
    public get isEmptyStateShown(): boolean {
        return this.isEmpty && !this.isNewApiKeyPopupShown && !this.hasSearchQuery;
    }

    /**
     * Fetches api keys page by clicked index.
     * @param index
     */
    public async onPageClick(index: number): Promise<void> {
        try {
            await this.$store.dispatch(FETCH, index);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }
    }

    /**
     * Used for sorting.
     * @param sortBy
     * @param sortDirection
     */
    public async onHeaderSectionClickCallback(sortBy: ApiKeyOrderBy, sortDirection: SortDirection): Promise<void> {
        await this.$store.dispatch(SET_SORT_BY, sortBy);
        await this.$store.dispatch(SET_SORT_DIRECTION, sortDirection);
        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        if (this.totalPageCount > 1) {
            this.$refs.pagination.resetPageIndex();
        }
    }

    /**
     * Sets api keys search query and then fetches depends on it.
     * @param query
     */
    public async onSearchQueryCallback(query: string): Promise<void> {
        await this.$store.dispatch(SET_SEARCH_QUERY, query);
        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        if (this.totalPageCount > 1) {
            this.$refs.pagination.resetPageIndex();
        }
    }
}
</script>

<style scoped lang="scss">
    .api-keys-area {
        position: relative;
        padding: 40px 30px 55px 30px;
        font-family: 'font_regular', sans-serif;

        &__title {
            font-family: 'font_bold', sans-serif;
            font-size: 32px;
            line-height: 39px;
            color: #263549;
            margin: 0;
        }

        .api-keys-header {
            width: 100%;
            position: relative;

            .blur-content {
                position: absolute;
                top: 100%;
                left: 0;
                background-color: #f5f6fa;
                width: 100%;
                height: 70vh;
                z-index: 100;
                opacity: 0.3;
            }

            .blur-search {
                position: absolute;
                bottom: 0;
                right: 0;
                width: 540px;
                height: 56px;
                z-index: 100;
                opacity: 0.3;
                background-color: #f5f6fa;
            }
        }

        .api-keys-items {
            position: relative;

            &__content {
                display: flex;
                flex-direction: column;
                width: 100%;
                justify-content: flex-start;
            }
        }
    }

    .empty-search-result-area {
        display: flex;
        align-items: center;
        justify-content: center;
        flex-direction: column;

        &__title {
            font-family: 'font_bold', sans-serif;
            font-size: 32px;
            line-height: 39px;
            margin-top: 104px;
        }

        &__image {
            margin-top: 40px;
        }
    }

    .pagination-area {
        margin-left: -25px;
        padding-bottom: 15px;
    }

    .header-default-state {
        display: flex;
        align-items: center;
    }

    .header-selected-api-keys {

        &__confirmation-label {
            font-family: 'font_medium', sans-serif;
            font-size: 14px;
            line-height: 28px;
        }

        &__buttons-area {
            display: flex;
            align-items: center;

            .deletion {
                margin-right: 12px;
            }
        }

        &__info-text {
            margin-left: 25px;
            line-height: 48px;
        }
    }

    .container.deletion {
        background-color: #ff4f4d;

        &.label {
            color: #fff;
        }

        &:hover {
            background-color: #de3e3d;
            box-shadow: none;
        }
    }

    .collapsed {
        margin-top: 0 !important;
        padding-top: 0 !important;
    }

    ::-webkit-scrollbar,
    ::-webkit-scrollbar-track,
    ::-webkit-scrollbar-thumb {
        width: 0;
    }
</style>
