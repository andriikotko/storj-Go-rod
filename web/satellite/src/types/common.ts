// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

export enum SortDirection {
    ASCENDING = 1,
    DESCENDING,
}
