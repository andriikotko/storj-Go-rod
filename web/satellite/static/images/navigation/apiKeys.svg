<svg class="svg" width="19" height="19" viewBox="0 0 19 19" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path class="navigation-svg-path" opacity="0.8" fill-rule="evenodd" clip-rule="evenodd" d="M19 0L18.8777 3.30167L17.3804 3.33154L17.3114 4.86798L15.548 5.164L15.5552 6.62419L14.0737 6.63827L13.8958 8.28354L12.3349 9.84422L13.2397 13.2214C13.3996 13.8182 13.229 14.4549 12.7921 14.8918L9.19044 18.4934C8.75356 18.9303 8.1168 19.1009 7.52001 18.941L2.60009 17.6227C2.00331 17.4628 1.53716 16.9967 1.37725 16.3999L0.0589658 11.48C-0.100943 10.8832 0.0696777 10.2464 0.506556 9.80956L4.10819 6.20793C4.54507 5.77105 5.18183 5.60043 5.77862 5.76034L9.15552 6.66484L15.6983 0.122284L19 0ZM6.52703 12.473C5.78414 11.7301 4.57967 11.7301 3.83678 12.473C3.09389 13.2159 3.09389 14.4203 3.83678 15.1632C4.57967 15.9061 5.78414 15.9061 6.52703 15.1632C7.26992 14.4203 7.26992 13.2159 6.52703 12.473Z" fill="#768394"/>
</svg>