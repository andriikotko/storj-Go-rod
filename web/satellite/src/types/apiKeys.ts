// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { SortDirection } from '@/types/common';

export type OnHeaderClickCallback = (sortBy: ApiKeyOrderBy, sortDirection: SortDirection) => Promise<void>;

/**
 * Exposes all apiKey-related functionality.
 */
export interface ApiKeysApi {
    /**
     * Fetch apiKeys
     *
     * @returns ApiKey[]
     * @throws Error
     */
    get(projectId: string, cursor: ApiKeyCursor): Promise<ApiKeysPage>;

    /**
     * Create new apiKey
     *
     * @returns ApiKey
     * @throws Error
     */
    create(projectId: string, name: string): Promise<ApiKey>;

    /**
     * Delete existing apiKey
     *
     * @returns null
     * @throws Error
     */
    delete(ids: string[]): Promise<void>;
}

/**
 * Holds api keys sorting parameters.
 */
export enum ApiKeyOrderBy {
    NAME = 1,
    CREATED_AT,
}

/**
 * ApiKeyCursor is a type, used to describe paged api keys list.
  */
export class ApiKeyCursor {
    public constructor(
        public search: string = '',
        public limit: number = 6,
        public page: number = 1,
        public order: ApiKeyOrderBy = ApiKeyOrderBy.NAME,
        public orderDirection: SortDirection = SortDirection.ASCENDING,
    ) {}
}

/**
 * ApiKeysPage is a type, used to describe paged api keys list.
 */
export class ApiKeysPage {
    public constructor(
        public apiKeys: ApiKey[] = [],
        public search: string = '',
        public order: ApiKeyOrderBy = ApiKeyOrderBy.NAME,
        public orderDirection: SortDirection = SortDirection.ASCENDING,
        public limit: number = 6,
        public pageCount: number = 0,
        public currentPage: number = 1,
        public totalCount: number = 0,
    ) {}
}

/**
 * ApiKey class holds info for ApiKeys entity.
 */
export class ApiKey {
    public isSelected: boolean;

    constructor(
        public id: string = '',
        public name: string = '',
        public createdAt: Date = new Date(),
        public secret: string = '',
    ) {
        this.isSelected = false;
    }

    /**
     * Returns created date as a local date string.
     */
    public localDate(): string {
        return this.createdAt.toLocaleDateString();
    }
}
