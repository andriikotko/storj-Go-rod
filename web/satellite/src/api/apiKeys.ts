// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { ApiKey, ApiKeyCursor, ApiKeysApi, ApiKeysPage } from '@/types/apiKeys';

/**
 * ApiKeysApiGql is a graphql implementation of ApiKeys API.
 * Exposes all apiKey-related functionality
 */
export class ApiKeysApiGql extends BaseGql implements ApiKeysApi {
    /**
     * Fetch apiKeys.
     *
     * @returns ApiKey
     * @throws Error
     */
    public async get(projectId: string, cursor: ApiKeyCursor): Promise<ApiKeysPage> {
        const query =
            `query($projectId: String!, $limit: Int!, $search: String!, $page: Int!, $order: Int!, $orderDirection: Int!) {
                project (
                    id: $projectId,
                ) {
                    apiKeys (
                        cursor: {
                            limit: $limit,
                            search: $search,
                            page: $page,
                            order: $order,
                            orderDirection: $orderDirection
                        }
                    ) {
                        apiKeys {
                            id,
                            name,
                            createdAt
                        }
                        search,
                        limit,
                        order,
                        pageCount,
                        currentPage,
                        totalCount
                    }
                }
            }`;

        const variables = {
            projectId: projectId,
            limit: cursor.limit,
            search: cursor.search,
            page: cursor.page,
            order: cursor.order,
            orderDirection: cursor.orderDirection,
        };

        const response = await this.query(query, variables);

        return this.getApiKeysPage(response.data.project.apiKeys);
    }

    /**
     * Used to create apiKey.
     *
     * @param projectId - stores current project id
     * @param name - name of apiKey that will be created
     * @returns ApiKey
     * @throws Error
     */
    public async create(projectId: string, name: string): Promise<ApiKey> {
        const query =
            `mutation($projectId: String!, $name: String!) {
                createAPIKey(
                    projectID: $projectId,
                    name: $name
                ) {
                    key,
                    keyInfo {
                        id,
                        name,
                        createdAt
                    }
                }
            }`;

        const variables = {
            projectId,
            name,
        };

        const response = await this.mutate(query, variables);
        const key: any = response.data.createAPIKey.keyInfo;
        const secret: string = response.data.createAPIKey.key;

        return new ApiKey(key.id, key.name, key.createdAt, secret);
    }

    /**
     * Used to delete apiKey.
     *
     * @param ids - ids of apiKeys that will be deleted
     * @throws Error
     */
    public async delete(ids: string[]): Promise<void> {
        const query =
            `mutation($id: [String!]!) {
                deleteAPIKeys(id: $id) {
                    id
                }
            }`;

        const variables = {
            id: ids,
        };

        const response = await this.mutate(query, variables);

        return response.data.deleteAPIKeys;
    }

    /**
     * Method for mapping api keys page from json to ApiKeysPage type.
     *
     * @param page anonymous object from json
     */
    private getApiKeysPage(page: any): ApiKeysPage {
        if (!page) {
            return new ApiKeysPage();
        }

        const apiKeysPage: ApiKeysPage = new ApiKeysPage();

        apiKeysPage.apiKeys = page.apiKeys.map(key => new ApiKey(
            key.id,
            key.name,
            new Date(key.createdAt),
            '',
        ));

        apiKeysPage.search = page.search;
        apiKeysPage.limit = page.limit;
        apiKeysPage.order = page.order;
        apiKeysPage.orderDirection = page.orderDirection;
        apiKeysPage.pageCount = page.pageCount;
        apiKeysPage.currentPage = page.currentPage;
        apiKeysPage.totalCount = page.totalCount;

        return apiKeysPage;
    }
}
