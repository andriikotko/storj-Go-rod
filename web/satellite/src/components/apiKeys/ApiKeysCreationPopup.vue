// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="new-api-key" v-if="isPopupShown">
        <h2 class="new-api-key__title">Name Your API Key</h2>
        <HeaderlessInput
            @setData="onChangeName"
            :error="errorMessage"
            placeholder="Enter API Key Name"
            class="full-input"
            width="100%"
        />
        <VButton
            class="next-button"
            label="Next >"
            width="128px"
            height="48px"
            :on-press="onNextClick"
        />
        <div class="new-api-key__close-cross-container" @click="onCloseClick">
            <CloseCrossIcon/>
        </div>
        <div class="blur-content"></div>
    </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

import HeaderlessInput from '@/components/common/HeaderlessInput.vue';
import VButton from '@/components/common/VButton.vue';

import CloseCrossIcon from '@/../static/images/common/closeCross.svg';

import { API_KEYS_ACTIONS } from '@/store/modules/apiKeys';
import { ApiKey } from '@/types/apiKeys';
import { SegmentEvent } from '@/utils/constants/analyticsEventNames';

const {
    CREATE,
    FETCH,
} = API_KEYS_ACTIONS;

@Component({
    components: {
        HeaderlessInput,
        VButton,
        CloseCrossIcon,
    },
})
export default class ApiKeysCreationPopup extends Vue {
    /**
     * Indicates if component should be rendered.
     */
    @Prop({default: false})
    private readonly isPopupShown: boolean;

    private name: string = '';
    private errorMessage: string = '';
    private isLoading: boolean = false;
    private key: string = '';

    private FIRST_PAGE = 1;

    public onChangeName(value: string): void {
        this.name = value.trim();
        this.errorMessage = '';
    }

    public onCloseClick(): void {
        this.onChangeName('');
        this.$emit('closePopup');
    }

    /**
     * Creates api key.
     */
    public async onNextClick(): Promise<void> {
        if (this.isLoading) {
            return;
        }

        if (!this.name) {
            this.errorMessage = 'API Key name can`t be empty';

            return;
        }

        this.isLoading = true;

        let createdApiKey: ApiKey;

        try {
            createdApiKey = await this.$store.dispatch(CREATE, this.name);
        } catch (error) {
            await this.$notify.error(error.message);
            this.isLoading = false;

            return;
        }

        await this.$notify.success('Successfully created new api key');
        this.key = createdApiKey.secret;
        this.isLoading = false;
        this.name = '';

        this.$segment.track(SegmentEvent.API_KEY_CREATED, {
            project_id: this.$store.getters.selectedProject.id,
        });

        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        this.$emit('closePopup');
        this.$emit('showCopyPopup', this.key);
    }
}
</script>

<style scoped lang="scss">
    .new-api-key {
        padding: 32px 58px 41px 40px;
        background-color: #fff;
        border-radius: 24px;
        margin-top: 29px;
        max-width: 93.5%;
        height: auto;
        position: relative;

        &__title {
            font-family: 'font_bold', sans-serif;
            font-size: 24px;
            line-height: 29px;
            margin-bottom: 26px;
        }

        .next-button {
            margin-top: 20px;
        }

        &__close-cross-container {
            display: flex;
            justify-content: center;
            align-items: center;
            position: absolute;
            right: 29px;
            top: 29px;
            height: 24px;
            width: 24px;
            cursor: pointer;

            &:hover .close-cross-svg-path {
                fill: #2683ff;
            }
        }

        .blur-content {
            position: absolute;
            top: 100%;
            left: 0;
            background-color: #f5f6fa;
            width: 100%;
            height: 70.5vh;
            z-index: 100;
            opacity: 0.3;
        }
    }
</style>
