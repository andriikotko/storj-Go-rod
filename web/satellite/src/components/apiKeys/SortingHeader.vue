// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="sort-header-container">
        <div class="sort-header-container__name-item" @click="onHeaderItemClick(ApiKeyOrderBy.NAME)">
            <p class="sort-header-container__name-item__title">Key Name</p>
            <VerticalArrows
                :is-active="areApiKeysSortedByName"
                :direction="getSortDirection"
            />
        </div>
        <div class="sort-header-container__date-item" @click="onHeaderItemClick(ApiKeyOrderBy.CREATED_AT)">
            <p class="sort-header-container__date-item__title creation-date">Created</p>
            <VerticalArrows
                :is-active="!areApiKeysSortedByName"
                :direction="getSortDirection"
            />
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

import VerticalArrows from '@/components/common/VerticalArrows.vue';

import { ApiKeyOrderBy, OnHeaderClickCallback } from '@/types/apiKeys';
import { SortDirection } from '@/types/common';

@Component({
    components: {
        VerticalArrows,
    },
})
export default class SortApiKeysHeader extends Vue {
    @Prop({default: () => new Promise(() => false)})
    private readonly onHeaderClickCallback: OnHeaderClickCallback;

    public ApiKeyOrderBy = ApiKeyOrderBy;

    public sortBy: ApiKeyOrderBy = ApiKeyOrderBy.NAME;
    public sortDirection: SortDirection = SortDirection.ASCENDING;

    /**
     * Used for arrow styling.
     */
    public get getSortDirection(): SortDirection {
        return this.sortDirection === SortDirection.DESCENDING ? SortDirection.ASCENDING : SortDirection.DESCENDING;
    }

    public get areApiKeysSortedByName(): boolean {
        return this.sortBy === ApiKeyOrderBy.NAME;
    }

    /**
     * Sets sorting kind if different from current.
     * If same, changes sort direction.
     * @param sortBy
     */
    public async onHeaderItemClick(sortBy: ApiKeyOrderBy): Promise<void> {
        if (this.sortBy !== sortBy) {
            this.sortBy = sortBy;
            this.sortDirection = SortDirection.ASCENDING;

            await this.onHeaderClickCallback(this.sortBy, this.sortDirection);

            return;
        }

        this.sortDirection = this.sortDirection === SortDirection.DESCENDING ?
            SortDirection.ASCENDING
            : SortDirection.DESCENDING;

        await this.onHeaderClickCallback(this.sortBy, this.sortDirection);
    }
}
</script>

<style scoped lang="scss">
    .sort-header-container {
        display: flex;
        width: 100%;
        height: 40px;
        background-color: rgba(255, 255, 255, 0.3);
        margin-top: 31px;

        &__name-item,
        &__date-item {
            width: 60%;
            display: flex;
            align-items: center;
            margin: 0;
            cursor: pointer;

            &__title {
                font-family: 'font_medium', sans-serif;
                font-size: 16px;
                margin: 0 0 0 80px;
                color: #2a2a32;
            }

            .creation-date {
                margin-left: 0;
            }
        }

        &__date-item {
            width: 40%;

            &__title {
                margin: 0;
            }
        }
    }
</style>
