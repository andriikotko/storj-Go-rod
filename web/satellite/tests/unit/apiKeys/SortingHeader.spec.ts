// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import sinon from 'sinon';

import SortingHeader from '@/components/apiKeys/SortingHeader.vue';

import { ApiKeyOrderBy } from '@/types/apiKeys';
import { SortDirection } from '@/types/common';
import { mount } from '@vue/test-utils';

describe('SortingHeader.vue', (): void => {
    it('should render correctly', (): void => {
        const wrapper = mount(SortingHeader);

        expect(wrapper).toMatchSnapshot();
    });

    it('should retrieve callback', (): void => {
        const onPressSpy = sinon.spy();

        const wrapper = mount(SortingHeader, {
            propsData: {
                onHeaderClickCallback: onPressSpy,
            },
        });
        wrapper.find('.sort-header-container__name-item').trigger('click');
        expect(onPressSpy.callCount).toBe(1);
    });

    it('should change sort direction', (): void => {
        const onPressSpy = sinon.spy();

        const wrapper = mount(SortingHeader, {
            propsData: {
                onHeaderClickCallback: onPressSpy,
            },
        });

        expect(wrapper.vm.$data.sortBy).toBe(ApiKeyOrderBy.NAME);
        expect(wrapper.vm.$data.sortDirection).toBe(SortDirection.ASCENDING);

        wrapper.find('.sort-header-container__name-item').trigger('click');
        expect(onPressSpy.callCount).toBe(1);

        expect(wrapper.vm.$data.sortBy).toBe(ApiKeyOrderBy.NAME);
        expect(wrapper.vm.$data.sortDirection).toBe(SortDirection.DESCENDING);
    });

    it('should change sort by value', (): void => {
        const onPressSpy = sinon.spy();

        const wrapper = mount(SortingHeader, {
            propsData: {
                onHeaderClickCallback: onPressSpy,
            },
        });

        expect(wrapper.vm.$data.sortBy).toBe(ApiKeyOrderBy.NAME);
        expect(wrapper.vm.$data.sortDirection).toBe(SortDirection.ASCENDING);

        wrapper.find('.sort-header-container__date-item').trigger('click');
        expect(onPressSpy.callCount).toBe(1);

        expect(wrapper.vm.$data.sortBy).toBe(ApiKeyOrderBy.CREATED_AT);
        expect(wrapper.vm.$data.sortDirection).toBe(SortDirection.ASCENDING);
    });
});
