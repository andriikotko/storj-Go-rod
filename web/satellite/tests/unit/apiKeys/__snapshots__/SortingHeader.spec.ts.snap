// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortingHeader.vue should render correctly 1`] = `
<div class="sort-header-container">
  <div class="sort-header-container__name-item">
    <p class="sort-header-container__name-item__title">Key Name</p>
    <div class="container"><svg width="9" height="6" viewBox="0 0 9 6" fill="none" xmlns="http://www.w3.org/2000/svg" class="">
        <path d="M4.73684 5.70565e-07L9 6L-9.53674e-07 6L4.73684 5.70565e-07Z" fill="#354049" class="arrow-svg-path"></path>
      </svg> <svg width="9" height="6" viewBox="0 0 9 6" fill="none" xmlns="http://www.w3.org/2000/svg" class="active">
        <path d="M4.26316 6L1.90735e-06 0L9 1.59559e-06L4.26316 6Z" fill="#354049" class="arrow-svg-path"></path>
      </svg></div>
  </div>
  <div class="sort-header-container__date-item">
    <p class="sort-header-container__date-item__title creation-date">Created</p>
    <div class="container"><svg width="9" height="6" viewBox="0 0 9 6" fill="none" xmlns="http://www.w3.org/2000/svg" class="">
        <path d="M4.73684 5.70565e-07L9 6L-9.53674e-07 6L4.73684 5.70565e-07Z" fill="#354049" class="arrow-svg-path"></path>
      </svg> <svg width="9" height="6" viewBox="0 0 9 6" fill="none" xmlns="http://www.w3.org/2000/svg" class="">
        <path d="M4.26316 6L1.90735e-06 0L9 1.59559e-06L4.26316 6Z" fill="#354049" class="arrow-svg-path"></path>
      </svg></div>
  </div>
</div>
`;
